rust/chains/hyperlane-cosmos/src/mailbox.rs

```diff
@@ -0,0 +1,378 @@
+            address: locator.address,
+            signer,
+            provider: Box::new(provider),
+        }
+    }
+}
+
+impl HyperlaneContract for CosmosMailbox {
+    fn address(&self) -> H256 {
+        self.address
+    }
+}
+
+impl HyperlaneChain for CosmosMailbox {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(CosmosProvider::new(self.domain.clone()))
+    }
+}
+
+impl Debug for CosmosMailbox {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        // Debug::fmt(&(self as &dyn HyperlaneContract), f)
+        todo!()
+    }
+}
+
+#[async_trait]
+impl Mailbox for CosmosMailbox {
+    #[instrument(level = "debug", err, ret, skip(self))]
+    async fn count(&self, lag: Option<NonZeroU64>) -> ChainResult<u32> {
+        let payload = mailbox::NonceRequest {
+            nonce: general::EmptyStruct {},
+        };
+
+        let data = self
+            .provider
+            .wasm_query(GeneralMailboxQuery { mailbox: payload }, lag)
+            .await;
+
+        if let Err(e) = data {
+            warn!("error: {:?}", e);
+            return Ok(0);
+        }
+
+        let response: mailbox::NonceResponse = serde_json::from_slice(&data?)?;
+        Ok(response.nonce)
+    }
+
+    #[instrument(level = "debug", err, ret, skip(self))]
+    async fn delivered(&self, id: H256) -> ChainResult<bool> {
+        let id = hex::encode(id);
+        let payload = mailbox::DeliveredRequest {
+            message_delivered: mailbox::DeliveredRequestInner { id },
+        };
+
+        let delivered = match self
+            .provider
+            .wasm_query(GeneralMailboxQuery { mailbox: payload }, None)
+            .await
+        {
+            Ok(v) => {
+                let response: mailbox::DeliveredResponse = serde_json::from_slice(&v)?;
+
+                response.delivered
+            }
+            Err(err) => {
+                warn!(
+                    "error while checking the message delivery status: {:?}",
+                    err
+                );
+
+                false
+            }
+        };
+
+        Ok(delivered)
+    }
+
+    #[instrument(err, ret, skip(self))]
+    async fn default_ism(&self) -> ChainResult<H256> {
+        let payload = mailbox::DefaultIsmRequest {
+            default_ism: general::EmptyStruct {},
+        };
+
+        let data = self
+            .provider
+            .wasm_query(GeneralMailboxQuery { mailbox: payload }, None)
+            .await?;
+        let response: mailbox::DefaultIsmResponse = serde_json::from_slice(&data)?;
+
+        // convert Hex to H256
+        let ism = H256::from_slice(&hex::decode(response.default_ism)?);
+        Ok(ism)
+    }
+
+    #[instrument(err, ret, skip(self))]
+    async fn recipient_ism(&self, recipient: H256) -> ChainResult<H256> {
+        let address = verify::digest_to_addr(recipient, &self.signer.prefix)?;
```

Comment:
we should get this from another source, i think it may be in the connection conf

Reply:
you mean get the prefix from another source? I removed the connection conf because it wasn't used, but why not use the signer for this?